### ISSUE TITLE
Rules for `copysign` are missing?

### ISSUE THREAD

**Issue by RainerHeintzmann** [JuliaDiff/ChainRules.jl#498]
`copysign` is used in `ChainRules.jl` in various places, which may have to do with adherence to the IEEE float standard.
Yet, `copysign` throws a, probably intended error:
```julia
julia> Zygote._pullback(copysign, 3.33, 4.4)[2](1)
ERROR: Non-differentiable function Core.Intrinsics.copysign_float
Stacktrace:
 [1] error(s::String)
   @ Base .\error.jl:33
 [2] macro expansion
   @ ~\.julia\packages\Zygote\TaBlo\src\compiler\interface2.jl:0 [inlined]
 [3] (::typeof(∂(λ)))(Δ::Int64)
   @ Zygote ~\.julia\packages\Zygote\TaBlo\src\compiler\interface2.jl:43
 [4] Pullback
   @ .\floatfuncs.jl:5 [inlined]
 [5] (::typeof(∂(copysign)))(Δ::Int64)
   @ Zygote ~\.julia\packages\Zygote\TaBlo\src\compiler\interface2.jl:0
 [6] top-level scope
   @ REPL[56]:1
```
, but the question is whether this is really required for the pullback rule. E.g. a function like `abs` works as expected but is also, strictly speaking, not differentiable at the origin. Maybe this could work:
```julia
@scalar_rule copysign(y, x) (sign(x)*y, 0)
```
or is there a fundamental reason that `copysign` should not have an allowed gradient, but `abs` should?

**Comment by willtebbutt** [comment]
Nope, this would be a good addition. If you have the time, a PR would be very welcome.

**Comment by RainerHeintzmann** [comment]
Sure, can do so. Maybe I should wait till my other pull request (`sincospi`) is merged and then add this?

**Comment by willtebbutt** [comment]
Sounds good 👍

**Comment by RainerHeintzmann** [comment]
Hmm, interestingly the `sign()` function yields a gradient of nothing:
```julia
julia> Zygote.gradient((x) -> sign(x), 1.2)
(nothing,)
```
should this not be `0`? Bug or feature?

**Comment by willtebbutt** [comment]
I think this is probably a Zygote thing. For historical reasons, Zygote represents zero tangents by `nothing`, whereas ChainRules uses either `0` or `ZeroTangent()`. I would _hope_ that if you called `rrule` directly you would obtain a `0`.

**Comment by RainerHeintzmann** [comment]
Yes, you are right:
```julia
julia> rrule(sign,-7.0)[2](1)
(NoTangent(), ZeroTangent())
```
indicating that `ZeroTangent()` may be more appropriate than `0` in my rule.
But then `sign` seems to also be defined for complex input, yielding a non-zero pullback:
```julia
julia> rrule(sign,-7.0)[2](1im)
(NoTangent(), 0.0 + 0.14285714285714285im)
```
Luckily `copysign()` is not allowed for complex inputs (except for `Unitful.AbstractQuantity`, as the first argument, which may be unintended), so maybe I can stick to the stated rule.

**Comment by RainerHeintzmann** [comment]
... The unitful version also (luckily) fails for `copysign`.

**Comment by willtebbutt** [comment]
Hmmm I'm slightly surprised that `sign` is giving you a `ZeroTangent` rather than `0` -- we usually go with `0` in scalar rules for performance-related reasons. @mcabbott @oxinabox @mzgubic any thoughts?

**Comment by mcabbott** [comment]
For `sign`, this doesn't look good, and is the reason to prefer `0.0`:
```
julia> gradient.(x -> x^2>1 ? sign(x) : x, (-3, -2, -1, 0, 1, 2, 3))  # super-crude tanh
((nothing,), (nothing,), (1,), (1,), (1,), (nothing,), (nothing,))

julia> gradient.(x -> ifelse(x^2>1, sign(x), x), (-3, -2, -1, 0, 1, 2, 3))  # weirdly ifelse fixes it?
((0,), (0,), (1,), (1,), (1,), (0,), (0,))

julia> gradient.(x -> clamp(x, -1, 1), (-3, -2, -1, 0, 1, 2, 3))  # better?
((0,), (0,), (1,), (1,), (1,), (0,), (0,))

julia> gradient.(x -> 0<x<1, (-3, -2, -1, 0, 1, 2, 3))  # this is always correctly categorical
((nothing,), (nothing,), (nothing,), (nothing,), (nothing,), (nothing,), (nothing,))
```
Unlike `signbit`, both `sign` and `copysign` preserve type, and thus are plausible things to use in stepwise functions like that:
```
julia> sign.((-33f0, -0f0, 0f0, 42f0))
(-1.0f0, -0.0f0, 0.0f0, 1.0f0)

julia> signbit.((-33f0, -0f0, 0f0, 42f0))
(true, true, false, false)

julia> copysign(2f0, big(-3.0))
-2.0f0
```
But I'm not sure the 2nd argument of `copysign` shouldn't be NoTangent. It does seem like it's the input to something like `x>0`. Maybe that's safe?

**Comment by RainerHeintzmann** [comment]
I see. So for type stability reason, 0 is clearly the better choice. Of course the `rrule` of `sign` can be easily fixed to return 0 instead of nothing as well. The clamp idea is probably not a good choice at least for real numbers, if you think about -0.5 as input  ;-)

**Comment by mcabbott** [comment]
No, I think I convinced myself of the opposite -- that the 2nd argument of `copysign` should be `NoTangent()`, like the 1st argument of `ifelse`. Unless someone can invent an example where this goes wrong.

I'm not sure why Zygote is giving `nothing` for `sign` above, BTW. The rule here doesn't seem to fix it:

https://github.com/JuliaDiff/ChainRules.jl/blob/e5c53380ac0537d37c2e7a380208964467482cd6/src/rulesets/Base/fastmath_able.jl#L201

although it does give a complex answer, and I don't see a more specific rule in here:

https://github.com/JuliaDiff/ChainRules.jl/blob/master/src/rulesets/Base/nondiff.jl

**Comment by oxinabox** [comment]
Zygote might have a `ZygoteRules.@no_grad sign`

I agree we should use NoTangent for the (co)tangent wrt the send argument.
Unless we find a real performance case where 0 is better.

Can always change in follow up.

(I am leaving this to someone else reviewing to approve and merge)

**Comment by mcabbott** [comment]
> Zygote might have a `@nograd sign`

I wondered this, but don't see anything: https://github.com/FluxML/Zygote.jl/search?q=sign&type=code

**Comment by mzgubic** [comment]
Closed by #502
